javascript/2331-evaluate-boolean-binary-tree.js

```diff
@@ -0,0 +1,28 @@
+/**
+ * Definition for a binary tree node.
+ * function TreeNode(val, left, right) {
+ *     this.val = (val===undefined ? 0 : val)
+ *     this.left = (left===undefined ? null : left)
+ *     this.right = (right===undefined ? null : right)
+ * }
+ */
+/**
+ * DFS | Tree-traversal
+ * Time O(n) | Space O(1)
+ * https://leetcode.com/problems/evaluate-boolean-binary-tree
+ * @param {TreeNode} root
+ * @return {boolean}
+ */
+var evaluateTree = function(root) {
+    return (dfs(root) && true) || false; 
```

Comment:
This is a bit odd, it would be cleaner if the  dfs should return the expected value

Reply:
How about this code?

```javascript
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * DFS | Tree-traversal
 * Time O(n) | Space O(1)
 * https://leetcode.com/problems/evaluate-boolean-binary-tree
 * @param {TreeNode} root
 * @return {boolean}
 */
var evaluateTree = function(root) {
    return dfs(root);
};

const dfs = (node) => {
    if(!node.left && !node.right && node.val) return true;
    if(!node.left && !node.right && !node.val) return false;

    const is2 = (node.val === 2);
    if(is2) return dfs(node.left) || dfs(node.right);
    
    const is3 = (node.val === 3);
    if(is3) return dfs(node.left) && dfs(node.right);

}
```

Reply:
@aakhtar3, any update? Here's the submission link: https://leetcode.com/problems/evaluate-boolean-binary-tree/submissions/1388673250/

Reply:
@aadil42 For lines 22 and 23, you can combine those two lines by using a single if statement to check if not node left and if not node right, then use a ternary to determine true or false depending on the node.val, see code snippet below:

**Original:**

```javascript
    if (!node.left && !node.right && node.val) return true;
    if (!node.left && !node.right && !node.val) return false;
```

**New:**
```javascript
    if (!node.left && !node.right) return (node.val) ? true : false;
```

Here's the submission with the tweak I made: https://leetcode.com/problems/evaluate-boolean-binary-tree/submissions/1426861839/

Reply:
@Ykhan799, I think the older version is easier to read the new puts more mental-strain as it has `if` statement inside `if` statement. But I updated the code with the new version. Please check Here's the submission link for it: https://leetcode.com/problems/evaluate-boolean-binary-tree/submissions/1431096463/

Reply:
@aadil42 Fair enough, you can revert the code to have the 2 if statements instead

Reply:
@Ykhan799, it's done.